src/conn.ts

```diff
@@ -23,106 +28,364 @@ export class ConnOptions {
     this.options = { ...new ConnOptions(), ...options };
     this.bot = createBot(botOptions) as any;
     this.bot.recipes = [];
+    this.receivingPclients = [];
+    this.pclients = [];
     this.write = this.bot._client.write.bind(this.bot._client);
     this.writeRaw = this.bot._client.writeRaw.bind(this.bot._client);
     this.writeChannel = this.bot._client.writeChannel.bind(this.bot._client);
-    this.bot._client.on('packet', (data, { name }, buffer) => {
-      //* relay packet to all connected clients
-      this.pclients.forEach((pclient) => {
-        if (!(pclient.toClientBlackList ?? this.options.toClientBlackList).includes(name)) pclient.writeRaw(buffer);
-      });
-      //* entity metadata tracking
-      if (data.metadata && data.entityId && this.bot.entities[data.entityId]) (this.bot.entities[data.entityId] as any).rawMetadata = data.metadata;
-
-      //* recipe tracking https://wiki.vg/index.php?title=Protocol&oldid=14204#Unlock_Recipes
-      switch (name) {
-        case 'unlock_recipes':
-          switch (data.action) {
-            case 0: //* initialize
-              this.bot.recipes = data.recipes1;
-              break;
-            case 1: //* add
-              this.bot.recipes = [...this.bot.recipes, ...data.recipes1];
-              break;
-            case 2: //* remove
-              this.bot.recipes = Array.from(
-                (data.recipes1 as number[]).reduce((recipes, recipe) => {
-                  recipes.delete(recipe);
-                  return recipes;
-                }, new Set(this.bot.recipes))
-              );
-              break;
+    if (options?.toClientMiddleware) this.toClientDefaultMiddleware = options.toClientMiddleware;
+    if (options?.toServerMiddleware) this.toServerDefaultMiddleware = options.toServerMiddleware;
+
+    this.bot._client.on('raw', this.onServerRaw.bind(this))
+  }
+
+  /**
+   * Called when the proxy bot receives a packet from the server. Forwards the packet to all attached and receiving clients taking
+   * attached middleware's into account.
+   * @param buffer Buffer
+   * @param meta 
+   * @returns 
+   */
+  async onServerRaw(buffer: Buffer, meta: PacketMeta) {
+    // @ts-ignore-error
+    const packetData = this.bot._client.deserializer.parsePacketBuffer(buffer).data.params
+    for (const pclient of this.receivingPclients) {
+      if (pclient.state !== states.PLAY || meta.state !== states.PLAY) { return }
+      // Build packet canceler function used by middleware
+      const cancel: packetCanceler = Object.assign((unCancel: boolean = false) => {
+        if (unCancel === true) {
+          cancel.isCanceled = false
+        } else {
+          cancel.isCanceled = true
+        }
+        update.isUpdated = true
+      }, { isCanceled: false })
+      const update: packetUpdater = Object.assign((unUpdate: boolean = false) => {
+        if (unUpdate === false) {
+          update.isUpdated = false
+        }
+        update.isUpdated = true
```

Comment:
this does not work... at all
```suggestion
        update.isUpdated = !!unUpdate;
```